### PR TITLE
Added solution to HW21

### PR DESCRIPTION
**Pull request by MichaelShevchenko**

### CODE REVIEW

**Review by viaviaj — APPROVED** [review]

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import core.basesyntax.dao.UserDao;
+import core.basesyntax.dao.UserDaoImpl;
+import core.basesyntax.model.Bet;
+import core.basesyntax.model.User;
+import java.util.Scanner;
+
+public class ConsoleHandler {
+    private static final String newUser = "n";
+    private static final String newBet = "b";
+    private static final String quitApp = "q";
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/controller/ConsoleHandler.java]: are these constants?

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                || command.equalsIgnoreCase(newBet)
+                || command.equalsIgnoreCase(quitApp));
+    }
+
+    private User newVisitor(Scanner scanner) {
+        System.out.println("Please, input some general information about you, "
+                            + "so we could accept and save your bets");
+        System.out.println("Enter your name and age");
+        String command = scanner.nextLine();
+        String[] userDate = command.split(" ");
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/controller/ConsoleHandler.java]:
```suggestion
        String[] userData = command.split(" ");
```

File: src/main/java/core/basesyntax/dao/UserDao.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.List;
+
+public interface UserDao {
+    boolean isLegalGambling(User user);
+
+    void saveAsGambler(User user);
+
+    List<User> getAllGamblers();
+
+    void makeBet(User user, Bet bet);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/dao/UserDao.java]: on dao layer there should be only methods related to storage manipulation
please move out `isLegalGambling()` and `makeBet()`

File: src/main/java/core/basesyntax/db/Storage.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax.db;
+
+import core.basesyntax.model.Bet;
+import core.basesyntax.model.User;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Bet> BETS = new ArrayList<>();
+    public static final List<User> LEGIT_GAMBLERS = new ArrayList<>();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/db/Storage.java]: though the modifiers are the same - these lists are not conceptually constants, there names should be lowercase

File: src/main/java/core/basesyntax/model/Bet.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package core.basesyntax.model;
+
+public class Bet {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/Bet.java:3]: add getters and setters

File: src/main/java/core/basesyntax/model/User.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package core.basesyntax.model;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class User {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/User.java:3]: here as well - add all getters and setters

File: src/main/java/core/basesyntax/model/User.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    private String name;
+    private int age;
+    private List<Bet> userBets;
+
+    public User(String name, int age) {
+        this.name = name;
+        this.age = age;
+        userBets = age >= 21 ? new ArrayList<>() : null;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/model/User.java]: if `userBets` is null, what happens when you call the method on it? (it happens in your app multiple times without a check)

let's not connect User with Bet please at this point
I would recommend to keep it simple and not to concentrate on business logic

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java (hunk 1 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/controller/ConsoleHandler.java]: done

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java (hunk 2 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/controller/ConsoleHandler.java]: done

File: src/main/java/core/basesyntax/dao/UserDao.java (hunk 3 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/dao/UserDao.java]: done

File: src/main/java/core/basesyntax/db/Storage.java (hunk 4 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/db/Storage.java]: done

File: src/main/java/core/basesyntax/model/Bet.java (hunk 5 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/model/Bet.java:3]: done

File: src/main/java/core/basesyntax/model/User.java (hunk 6 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/model/User.java:3]: done

File: src/main/java/core/basesyntax/model/User.java (hunk 7 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/model/User.java]: done

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        return command;
+    }
+
+    private boolean isValidCommand(String command) {
+        return (command.equalsIgnoreCase(NEW_USER)
+                || command.equalsIgnoreCase(NEW_BET)
+                || command.equalsIgnoreCase(QUIT_APP));
+    }
+
+    private User newVisitor(Scanner scanner) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/controller/ConsoleHandler.java]:
```suggestion
    private User createNewVisitor(Scanner scanner) {
```

File: src/main/java/core/basesyntax/controller/ConsoleHandler.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        } catch (NumberFormatException | ArrayIndexOutOfBoundsException message) {
+            System.out.println("Please, input appropriate data");
+        }
+        System.out.println(visitor == null
+                ? "Unfortunately, provided information is not applicable for correct processing"
+                : "Your data has been successfully processed");
+        return visitor;
+    }
+
+    private String newBet(Scanner scanner) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/controller/ConsoleHandler.java]:
```suggestion
    private String createNewBet(Scanner scanner) {
```
method names should be verbs

File: src/main/java/core/basesyntax/db/Storage.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax.db;
+
+import core.basesyntax.model.Bet;
+import core.basesyntax.model.User;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Storage {
+    public static final List<Bet> bets = new ArrayList<>();
+    public static final List<User> legit_gamblers = new ArrayList<>();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/db/Storage.java]: how do we write variable names according to the naming convention?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Comment by MichaelShevchenko** [comment]
начал делать вторую часть домашки до того как эпрувнули первую, поэтому тут по идее ещё отразятся изменения обусловленные ревью до эпрува

File: src/main/java/core/basesyntax/lib/Injector.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            if (field.getAnnotation(Inject.class) != null
+                    && field.getType().equals(UserDao.class)) {
+                field.setAccessible(true);
+                field.set(instance, Factory.getUserDao());
+            }
+            if (Factory.getUserDao().getClass().getAnnotation(Dao.class) == null
+                    || Factory.getBetDao().getClass().getAnnotation(Dao.class) == null) {
+                throw new DaoAnnotationException("There is no proper annotation "
+                        + "in Dao implementation class");
+            }
```
Comment (dmytro-hryhoruk) [inline src/main/java/core/basesyntax/lib/Injector.java]: let's make this check before setting instance

**Review by dmytro-hryhoruk — COMMENTED** [review]
Good job!Let's refactor this PR now

File: src/main/java/core/basesyntax/lib/Injector.java (hunk 11 above)
Comment (MichaelShevchenko) [reply src/main/java/core/basesyntax/lib/Injector.java]: done

**Pull request closed without merging**